go-controller/pkg/libovsdbops/vswitchd.go

```diff
@@ -0,0 +1,279 @@
+// given predicate
+func FindInterfacesWithPredicate(vsClient libovsdbclient.Client, p InterfacePredicate) ([]*vswitchdb.Interface, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), types.OVSDBTimeout)
+	defer cancel()
+	found := []*vswitchdb.Interface{}
+	err := vsClient.WhereCache(p).List(ctx, &found)
+	return found, err
+}
+
+// FindInterfaceByName looks up an Interface from the cache by name
+func FindInterfaceByName(vsClient libovsdbclient.Client, ifaceName string) (*vswitchdb.Interface, error) {
+	found := []*vswitchdb.Interface{}
+	opModel := operationModel{
+		Model:          &vswitchdb.Interface{Name: ifaceName},
+		ExistingResult: &found,
+		ErrNotFound:    true,
+		BulkOp:         false,
+	}
+
+	m := newModelClient(vsClient)
+	if err := m.Lookup(opModel); err != nil {
+		return nil, fmt.Errorf("error looking up Interface %q: %w", ifaceName, err)
+	}
+
+	return found[0], nil
+}
+
+// FindPortByName looks up a Port from the cache by name
+func FindPortByName(vsClient libovsdbclient.Client, portName string) (*vswitchdb.Port, error) {
+	found := []*vswitchdb.Port{}
+	opModel := operationModel{
+		Model:          &vswitchdb.Port{Name: portName},
+		ExistingResult: &found,
+		ErrNotFound:    true,
+		BulkOp:         false,
+	}
+
+	m := newModelClient(vsClient)
+	if err := m.Lookup(opModel); err != nil {
+		return nil, fmt.Errorf("error looking up Port %q: %w", portName, err)
+	}
+
+	return found[0], nil
+}
+
+// CreateQoS deletes QoS records with a "sandbox" ExternalID that
+// matches the given sandbox ID
+func CreateQoS(vsClient libovsdbclient.Client, sandboxID string, maxRateBPS int64) (*vswitchdb.QoS, error) {
+	qos := &vswitchdb.QoS{
+		Type: "linux-htb",
+		ExternalIDs: map[string]string{
+			"sandbox": sandboxID,
+		},
+		OtherConfig: map[string]string{
+			"max-rate": fmt.Sprintf("%d", maxRateBPS),
+		},
+	}
+	opModel := operationModel{
+		Model:       qos,
+		ErrNotFound: false,
+		BulkOp:      true,
+	}
+
+	m := newModelClient(vsClient)
+	if _, err := m.CreateOrUpdate(opModel); err != nil {
+		return nil, err
+	}
+	return qos, nil
+}
+
+// DeleteQoSBySandboxID deletes QoS records with a "sandbox" ExternalID that
+// matches the given sandbox ID
+func DeleteQoSBySandboxID(vsClient libovsdbclient.Client, sandboxID string) error {
+	opModel := operationModel{
+		Model: &vswitchdb.QoS{},
+		ModelPredicate: func(item *vswitchdb.QoS) bool {
+			foundID, ok := item.ExternalIDs["sandbox"]
+			return ok && foundID == sandboxID
+		},
+		ErrNotFound: false,
+		BulkOp:      true,
+	}
+
+	m := newModelClient(vsClient)
+	return m.Delete(opModel)
+}
+
+// CreateOrUpdatePortAndAddToBridge creates or updates the provided Interface
+// Interfae template, provided Port template, and adds the Port to the given bridge
+func CreateOrUpdatePortAndAddToBridge(vsClient libovsdbclient.Client, bridgeUUID string, portTemplate *vswitchdb.Port, ifaceTemplate *vswitchdb.Interface) error {
```

Comment:
nit: 

By convention, these type of methods would have this signature
```
func CreateOrUpdateInterfacePortOnBridge(nbClient libovsdbclient.Client, bridge *vswitchdb.Bridge, port *vswitchdb.Port, interface *vswitchdb.Interface) error
```

I guess that we don't care whether we are provided a bridge UUID or a bridge name.

Perhaps we could panic if ifaceTemplate.Name != portTemplate.Name